src/cargo/util/cache_lock.rs

```diff
@@ -0,0 +1,463 @@
+//! Support for locking the package and index caches.
+//!
+//! This implements locking on the package and index caches (source files,
+//! `.crate` files, and index caches) to coordinate when multiple cargos are
+//! running at the same time. There are three styles of locks:
+//!
+//! * [`CacheLockMode::DownloadExclusive`] -- This is an exclusive lock
+//!   acquired while downloading packages and doing resolution.
+//! * [`CacheLockMode::Shared`] -- This is a shared lock acquired while a
+//!   build is running. In other words, whenever cargo just needs to read from
+//!   the cache, it should hold this lock. This is here to ensure that no
+//!   cargos are trying to read the source caches when cache garbage
+//!   collection runs.
+//! * [`CacheLockMode::MutateExclusive`] -- This is an exclusive lock acquired
+//!   whenever needing to modify existing source files (for example, with
+//!   cache garbage collection). This is acquired to make sure that no other
+//!   cargo is reading from the cache.
+//!
+//! Importantly, a `DownloadExclusive` lock does *not* interfere with a
+//! `Shared` lock. The download process generally does not modify source
+//! files, so other cargos should be able to safely proceed in reading source
+//! files[^1].
```

Comment:
I feel like most of this should be on the enum because that is where people will look (for what might not already be) and we should avoid duplicating such detailed documentation.

Reply:
Looks like these descriptions are more user facing / policy driven while the enum has more technical details

Reply:
Can you say what on the enum seems unclear? Or is the concern that this lists the modes here? The intent is the module docs provide a global description of how everything in the module works and fits together. It seems hard to me to do that without first introducing what possible lock modes exist. This introduction is about as brief as possible needed to give a high-level understanding of what the modes are, and then the enum itself has detailed documentation on each mode. I could remove the list here and say "go read the `CacheLockMode` docs first, and then come back", but that seems a little awkward to me.

Or are you asking to move the entire module docs to the enum? That doesn't seem quite right to me because this documentation is covering more than just the modes, but instead how everything here fits together (modes, locker, files, invariants the locker needs to hold, etc.).

Reply:
The following is in the mod comment and is clearer than what is on the enum

> acquired while downloading packages and doing resolution.

"Oh, resolution, thats why `cargo add` and `cargo remove` need it".

One challenge with this doc comment is its serving double-duty.  It covers both user-facing information and implementation details.  Looking at it, I could see also moving the "old cargo versions" part down to the enum as well.

Maybe a different angle: how much of these details should even be a doc  comment on the mod vs a doc or regular comment on `CacheState`?

Reply:
I think it is useful to have an overview of how everything fits together in one place. There are important aspects of the high-level picture that I think are useful to know, as well as some of the lower-level details (since these aren't perfect abstractions). If everything was split up, then one would need to go hunting through multiple places to get a understanding of how the pieces fit together and what the different inter-playing aspects are.  Things like the old cargo versions isn't relevant to choosing one of the enum variants, so I'm not sure I understand why it would be moved there. The intent is the enum docs are specifically for choosing which enum to use, not to explain how the system works overall (which I think would distract from the act of trying to pick a variant).

I have added the "and resolution" part to the enum. Otherwise, I think the enum docs currently explain how to choose which variant to use.

Is this an important concern, or a non-blocking nit?

Reply:
> I have added the "and resolution" part to the enum. Otherwise, I think the enum docs currently explain how to choose which variant to use.

Only `DownloadExclusive` has guidance on when to use it.  Ways to improve it (part of my "move things down").  The top-level docs talk about builds for `shared` and garbage collection for `garbage collection`.  Having those in would at least provide some context to what people are reading when reading the documentation for the variants.

> Is this an important concern, or a non-blocking nit?

With how complicated this is, I would consider having actionable documentation to be blocking.  Refinement past that would not be.

Reply:
I extended the documentation with more guidance on how to use the locks and the different modes.